README.md
# Storybook Puppeteer

[![NPM version][npm-image]][npm-url]
[![License][license-image]][license-url]
[![Commitizen friendly][commitizen-image]][commitizen-url]

`storybook-puppeteer` is a CLI to run end-to-end tests for storybook stories.  

## Alternatives

[Jest](https://github.com/facebook/jest) gives you full control to test your components.  
If you are using a modern stack which allows using jest please concider using Jest directly or one of the following modules instead:

- [https://www.npmjs.com/package/@storybook/addon-storyshots](@storybook/addon-storyshots)
- [https://github.com/smooth-code/jest-puppeteer](jest-puppeteer)

## Getting started

`storybook-puppeteer` has 0 dependencies however it relies on puppeteer to be installed as a peer dependency.  
If puppeteer is not installed it will be installed during execution.

NPM 

```bash
npm i --save-dev storybook-puppeteer
npx storybook-puppeteer
```

```bash
yarn add --dev storybook-puppeteer
yarn storybook-puppeteer
```

![cli output demo](https://raw.githubusercontent.com/jantimon/storybook-puppeteer/master/storybook.png)


## Options

```
Options:
  --help                 Show help                                     [boolean]
  --version              Show version number                           [boolean]
  -t, --timeout          Timeout for page load or selector lookup
  -d, --waitDuration     The duration after the page load to look for runtime
                         errors in miliseconds
  -s, --waitForSelector  Assert that a given selector is present in dom
  -w, --maxWorkers       Specifies the maximum number of workers the worker-pool
                         will spawn for running tests
  -u, --url              The base url of the storybook instance e.g.
                         http://localhost:6009/
  -b, --bail             Stop on first error
  -e, --exclude          Exclude a story e.g. -e "base-intro" or -e "demo-*".
                         Can be used multiple times                      [array]
  -o, --only             Run only the stories e.g. -o "base-intro" or -o
                         "demo-*". Can be used multiple times            [array]
  --serveDirectory       Serves the directory of a static storybook build
  --servePort            The port used for serveDirectory
  --autoInstall          Install puppeteer if missing - true by default
```

## License

[MIT license](http://opensource.org/licenses/MIT)

[npm-image]: https://badge.fury.io/js/storybook-puppeteer.svg
[npm-url]: https://npmjs.org/package/storybook-puppeteer
[license-image]: https://img.shields.io/badge/license-MIT-green.svg
[license-url]: http://opensource.org/licenses/MIT
[commitizen-image]: https://img.shields.io/badge/commitizen-friendly-brightgreen.svg
[commitizen-url]: http://commitizen.github.io/cz-cli/

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.8.0](https://github.com/jantimon/storybook-puppeteer/compare/v1.7.0...v1.8.0) (2020-10-07)


### Features

* add support for storybook 6 ([c497ba8](https://github.com/jantimon/storybook-puppeteer/commit/c497ba80f9adfc5e84ef6870becde0ca1f5c7a05))

## [1.7.0](https://github.com/jantimon/storybook-puppeteer/compare/v1.6.1...v1.7.0) (2020-08-05)


### Features

* use emotjis for ✅ and ❌ ([80f960f](https://github.com/jantimon/storybook-puppeteer/commit/80f960f1f1e25eac3521e1accc7b9b0604c98290))

### [1.6.1](https://github.com/jantimon/storybook-puppeteer/compare/v1.6.0...v1.6.1) (2020-08-05)


### Bug Fixes

* show errors of retries ([b14ad30](https://github.com/jantimon/storybook-puppeteer/commit/b14ad30bc26bbacfebdb774037949edb0b829bc0))

## [1.6.0](https://github.com/jantimon/storybook-puppeteer/compare/v1.5.1...v1.6.0) (2020-08-04)


### Features

* retry failed stories ([cf047b5](https://github.com/jantimon/storybook-puppeteer/commit/cf047b508e2dc50b32576b9e2c72476a872efc25))

### [1.5.1](https://github.com/jantimon/storybook-puppeteer/compare/v1.5.0...v1.5.1) (2020-07-31)


### Bug Fixes

* hide error messages during puppeteer avaiability check ([f1d6864](https://github.com/jantimon/storybook-puppeteer/commit/f1d6864c5c7bb75035464e0674cd4cd06c1eb25c))

## [1.5.0](https://github.com/jantimon/storybook-puppeteer/compare/v1.4.0...v1.5.0) (2020-07-31)


### Features

* remove puppeteer as optional dependency ([3a5aada](https://github.com/jantimon/storybook-puppeteer/commit/3a5aada61f9a08b70a1c2837c3103492c9b7bbc5))

## [1.4.0](https://github.com/jantimon/storybook-puppeteer/compare/v1.3.1...v1.4.0) (2020-07-16)


### Features

* output baseUrl ([fd5e143](https://github.com/jantimon/storybook-puppeteer/commit/fd5e143d20ef47982f8a4aebda798264d927dea4))

### [1.3.1](https://github.com/jantimon/storybook-puppeteer/compare/v1.3.0...v1.3.1) (2020-07-16)


### Bug Fixes

* show correct error message for default base url ([182f36f](https://github.com/jantimon/storybook-puppeteer/commit/182f36f7b0663ed68a4b3574cb88fbc7d2104395))

## [1.3.0](https://github.com/jantimon/storybook-puppeteer/compare/v1.2.0...v1.3.0) (2020-07-14)


### Features

* add puppeteer autoinstall ([fb08ab0](https://github.com/jantimon/storybook-puppeteer/commit/fb08ab024684a4efd9788bb4ddefc66157c5a4dc))

## [1.2.0](https://github.com/jantimon/storybook-puppeteer/compare/v1.1.0...v1.2.0) (2020-07-13)


### Features

* launch headless ([cee37ed](https://github.com/jantimon/storybook-puppeteer/commit/cee37edf79d54d4f910dfa088520123d784567d8))

## [1.1.0](https://github.com/jantimon/storybook-puppeteer/compare/v1.0.0...v1.1.0) (2020-07-12)


### Features

* add support for serving static storybook builds ([11c5896](https://github.com/jantimon/storybook-puppeteer/commit/11c5896abd1af5e011a81a80a8b0069fabb1c723))

## 1.0.0 (2020-07-12)


src/lib.ts
import { Browser, Page } from "puppeteer";

const sleep = (duration: number) =>
  new Promise((resolve) => setTimeout(resolve, duration));

export const DEFAULT_BASE_URL = "http://localhost:6009/";

declare var __STORYBOOK_STORY_STORE__: // Storybook 5
| {
      getStoriesForManager(): {
        [key: string]: { id: string; kind: string; name: string };
      };
    }
  // Storybook 6
  | {
      getStoriesJsonData(): {
        stories: {
          [key: string]: { id: string; kind: string; name: string };
        };
      };
    };

export async function getStories(
  page: Page,
  baseUrl = DEFAULT_BASE_URL
): Promise<{ id: string; kind: string; name: string }[]> {
  await page.goto(`${baseUrl}iframe.html`);
  // Wait for javascript execution (once storybook rendered)
  await waitForGlobaLVariable(page, "__STORYBOOK_STORY_STORE__");
  // Return stories
  return await page.evaluate(() => {
    const stories =
      "getStoriesJsonData" in __STORYBOOK_STORY_STORE__
        ? __STORYBOOK_STORY_STORE__.getStoriesJsonData().stories
        : __STORYBOOK_STORY_STORE__.getStoriesForManager();
    return Object.keys(stories).map((storyId) => {
      const { id, kind, name } = stories[storyId];
      return { id, kind, name };
    });
  });
}

export async function assertStoryHasNoErrors(
  page: Page,
  id: string,
  baseUrl = DEFAULT_BASE_URL,
  timeout = 5000,
  waitForDuration = 1000,
  waitForSelector?: string
) {
  const errors: Array<any> = [];
  const handleError = (err: any) => {
    errors.push(err);
  };
  page.on("pageerror", handleError);

  try {
    await page.goto(`${baseUrl}iframe.html?id=${id}`, {
      waitUntil: "load",
      timeout,
    });
    await waitForGlobaLVariable(page, "__STORYBOOK_STORY_STORE__");
    await sleep(waitForDuration);
    if (waitForSelector) {
      await page.waitForSelector(waitForSelector, {
        timeout,
      });
    }
  } catch (e) {
    page.off("pageerror", handleError);
    if (errors.length) {
      throw errors[0];
    }
    throw e;
  }
  page.off("pageerror", handleError);
  if (errors.length) {
    throw errors[0];
  }
}

async function waitForGlobaLVariable(
  page: Page,
  name: string,
  timeout: number = 5000
) {
  return new Promise(async (resolve, reject) => {
    const timer = setTimeout(
      () => reject(new Error(`Global Variable "${name}" timed out`)),
      timeout
    );
    while (true) {
      if (await page.evaluate(`typeof ${name} !== 'undefined'`)) {
        clearTimeout(timer);
        resolve();
        return;
      }
      await sleep(100);
    }
  });
}


webpack.config.js
const webpack = require('webpack');
const TsConfigWebpackPlugin = require('ts-config-webpack-plugin');

module.exports = {
  devtool: false,
  target: 'node',
  output: {
    path: __dirname + '/bin'
  },
  optimization: {
    minimize: false
  },
  plugins: [
    // Multi threading typescript loader configuration with caching for .ts and .tsx files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/ts-config-webpack-plugin/config
    new TsConfigWebpackPlugin(),
    new webpack.BannerPlugin({
      raw: true,
      banner: '#!/usr/bin/env node'
    })
  ],
  externals: {
    // Keep PeerDependencies out of the bundle
    puppeteer: 'require("puppeteer")',
    npm: 'require("npm")',
  },
  node: {
    __dirname: false
  },
  stats: {
    // Ignore warnings due to yarg's dynamic module loading
    warningsFilter: [/node_modules\/yargs/, /node_modules\/express/]
  },
};


src/index.ts
import express from "express";
import yargs = require("yargs");
import globToRegExp from "glob-to-regexp";
import chalk from "chalk";
import { Page } from "puppeteer";
import path from 'path';
import { execSync } from 'child_process';
import "express";
import { getStories, assertStoryHasNoErrors, DEFAULT_BASE_URL } from "./lib";
const checknark = chalk.greenBright("✅");
const fail = chalk.red("❌");

function getCliOptions() {
  // Parse command line options
  const {
    d: waitDuration,
    w: maxWorkers,
    b: bail,
    u: baseUrl,
    s: waitForSelector,
    t: timeout,
    e: exclude,
    o: only,
    autoInstall,
    serveDirectory,
    servePort,
  } = yargs
    .usage("Usage: $0 [options]")
    .alias("t", "timeout")
    .nargs("t", 1)
    .describe("t", "Timeout for page load or selector lookup")
    .alias("d", "waitDuration")
    .nargs("d", 1)
    .describe(
      "d",
      "The duration after the page load to look for runtime errors in miliseconds"
    )
    .alias("s", "waitForSelector")
    .nargs("s", 1)
    .describe("s", "Assert that a given selector is present in dom")
    .alias("w", "maxWorkers")
    .nargs("w", 1)
    .describe(
      "w",
      "Specifies the maximum number of workers the worker-pool will spawn for running tests"
    )
    .alias("u", "url")
    .nargs("u", 1)
    .describe(
      "u",
      "The base url of the storybook instance e.g. http://localhost:6009/"
    )
    .alias("b", "bail")
    .describe("b", "Stop on first error")
    .alias("e", "exclude")
    .array("e")
    .describe(
      "e",
      'Exclude a story e.g. -e "base-intro" or -e "demo-*". Can be used multiple times'
    )
    .alias("o", "only")
    .array("o")
    .describe(
      "o",
      'Run only the stories e.g. -o "base-intro" or -o "demo-*". Can be used multiple times'
    )
    .describe(
      "serveDirectory",
      "Serves the directory of a static storybook build"
    )
    .describe("servePort", "The port used for serveDirectory")
    .nargs("autoInstall", 1)
    .describe(
      "autoInstall",
      "Install puppeteer if missing - true by default"
    )
    .argv as Partial<{
    d: string;
    w: string;
    b: boolean;
    u: string;
    s: string;
    t: string;
    e: string[];
    o: string[];
    autoInstall: string;
    serveDirectory: string;
    servePort: string;
  }>;

  const excludeRegularExpressions = (exclude || []).map((exclusion) =>
    globToRegExp(String(exclusion), { extended: true })
  );
  const onlyRegularExpressions = (only || []).map((inclusion) =>
    globToRegExp(String(inclusion), { extended: true })
  );

  const toNumer = (val: string | undefined) =>
    typeof val === "string" && val !== "" ? Number(val) : undefined;

  const cliOptions = {
    waitDuration: toNumer(waitDuration) || 200,
    maxWorkers: toNumer(maxWorkers) || 5,
    bail: !!bail,
    baseUrl: baseUrl || DEFAULT_BASE_URL,
    waitForSelector: waitForSelector || "#root *",
    timeout: toNumer(timeout) || 5000,
    excludeRegularExpressions,
    onlyRegularExpressions,
    serveDirectory,
    servePort: toNumer(servePort) || 6011,
    autoInstall: autoInstall !== 'false'
  };

  // Proivde a different baseUrl default value if serveDirectory option is set
  if (serveDirectory && !baseUrl) {
    cliOptions.baseUrl = `http://localhost:${cliOptions.servePort}/`;
  }

  return cliOptions;
}

const cliOptions = getCliOptions();

async function main(pages: Page[]) {
  console.log("🚀 Launching " + require("../package.json").name);

  let tearDownServer = () => {};
  if (cliOptions.serveDirectory) {
    console.log("🌐 Starting server on port " + chalk.blue(cliOptions.servePort));
    tearDownServer = await launchExpressServer(cliOptions.serveDirectory, cliOptions.servePort);
  }

  console.log("   Test stories on " + chalk.underline.blue(cliOptions.baseUrl));
  const storiesPromise = getStories(pages[0], cliOptions.baseUrl);
  // Output launch errors
  try {
    await storiesPromise;
  } catch (e) {
    console.log(`${fail} Could not find any stories$ for ${chalk.grey(cliOptions.baseUrl)}. Is your storybook server running?`);
    console.log(chalk.red(e.message));
    process.exit(1);
  }
  const allStories = await storiesPromise;

  // exclude / include stories based on cli arguments
  const stories = allStories.filter(
    (story) =>
      !cliOptions.excludeRegularExpressions.some((regExp) =>
        regExp.test(story.id)
      ) &&
      (cliOptions.onlyRegularExpressions.length === 0 ||
        cliOptions.onlyRegularExpressions.some((regExp) =>
          regExp.test(story.id)
        ))
  );

  console.log(
    ` ${
      stories.length !== 0 ? chalk.greenBright(stories.length) : chalk.red(0)
    } ${stories.length === 1 ? "story" : "stories"} found.`
  );

  console.log(` starting up ${pages.length} workers`);
  const errors = await testStories(pages, stories);
  tearDownServer();
  
  if (errors) {
    console.log(
      chalk.red(
        `${errors} of ${stories.length} ${
          stories.length !== 1 ? "stories" : "story"
        } failed.`
      )
    );
    process.exit(1);
  } else {
    console.log(
      chalk.green(
        `${stories.length} ${
          stories.length !== 1 ? "stories" : "story"
        } succeeded.`
      )
    );
  }
}

async function testStories(
  pages: Page[],
  stories: {
    id: string;
    kind: string;
    name: string;
  }[]
) {
  const storyStack = [...stories];

  let errors = 0;
  async function takeNext(page: Page): Promise<void> {
    const story = storyStack.shift();
    const start = Date.now();
    if (!story) {
      return;
    }
    let testError: Error | undefined;
    try {
      await testStory(page, story.id);
    } catch (err) {
      testError = err;
    }
    if (testError) {
      errors++;
      console.log(` ${fail} ${story.kind} - ${story.name}`);
      console.log(
        `   ${chalk.underline.blue(
          cliOptions.baseUrl + "iframe.html?id=" + story.id
        )}`
      );
      console.log("   " + testError.toString());
      console.log("\n");
      if (cliOptions.bail) {
        process.exit(1);
      }
    } else {
      console.log(
        ` ${checknark} ${story.id} (${story.kind} - ${story.name}) in ${(
          Math.round(Date.now() - start) / 1000
        ).toFixed(3)}s`
      );
    }
    return takeNext(page);
  }

  // Execute in parallel:
  await Promise.all(pages.map((page) => takeNext(page)));

  return errors;
}

async function testStory(
  page: Page,
  id: string,
  retries: number = 3
) {
  try {
    await assertStoryHasNoErrors(
      page,
      id,
      cliOptions.baseUrl,
      cliOptions.timeout,
      cliOptions.waitDuration,
      cliOptions.waitForSelector
    );
  } catch(e) {
    // Ignore the error and retry to test the story
    if (retries) {
      await testStory(page, id, retries -1);
    } else {
      // Rethrow the error if no more retries are possible
      throw e;
    }
  }
}

function launchExpressServer(directory: string, port: number) {
  return new Promise<() => void>((resolve, reject) => {
    const app = express();
    app.use(express.static(directory));
    let server = app.listen(port, (error: any) => error ? reject(error) : resolve(() => server.close()));
  })
}

function isPuppeteerInstalled() {
  try {
    execSync('node -e \'require("puppeteer")\'', {cwd: __dirname, stdio: ['pipe', 'pipe', 'ignore']});
  } catch(e) {
    return false;
  }
  return true;
}

function installPuppeteer() {
  console.log('🚚 install puppeteer')
  const binaryLocationParts = __dirname.split(path.sep);
  const installLocation = (binaryLocationParts[binaryLocationParts.length - 2] === 'node_modules' ? binaryLocationParts.slice(0, -3) : binaryLocationParts.slice(0, -1)).join(path.sep);
  console.log(execSync('npm install puppeteer --no-save', {
    cwd: installLocation
  }).toString());
}


// Start
if (!isPuppeteerInstalled() && cliOptions.autoInstall) {
  installPuppeteer();
}
const puppeteer = __non_webpack_require__('puppeteer') as typeof import('puppeteer');
puppeteer.launch({
  headless: true,
  args: [
    '--no-sandbox',
    '--disable-setuid-sandbox',
  ]
}).then(async (browser) => {
  // Generate a tab per worker
  // Use min 1 worker and max 15 workers
  const pages = await Promise.all(
    Array(Math.min(Math.max(1, cliOptions.maxWorkers || 1), 15))
      .fill("")
      .map(() => browser.newPage())
  );
  try {
    await main(pages);
  } catch (e) {
    await browser.close();
    throw e;
  }
  await browser.close();
  process.exit();
});